File: 138. Copy List with Random Pointer.js
/**
 * Definition for singly-linked list with a random pointer.
 * function RandomListNode(label) {
 *     this.label = label;
 *     this.next = this.random = null;
 * }
 */

/**
 * @param {RandomListNode} head
 * @return {RandomListNode}
 */
var copyRandomList = function(head) {
    if ( head == null ) {
        return null;
    }

    var map = new Map();
    var dummyNode = new RandomListNode(0);
    var pre = dummyNode;
    var newNode;

    while ( head != null ) {
        if ( map.has(head) ) {
            newNode = map.get(head);
        } else {
            newNode = new RandomListNode( head.label );
            map.set( head, newNode );
        }
        pre.next = newNode;

        if ( head.random != null ) {
            if ( map.has( head.random ) ) {
                newNode.random = map.get(head.random);
            } else {
                newNode.random = new RandomListNode( head.random.label );
                map.set(head.random, newNode.random);
            }
        }

        pre = newNode;
        head = head.next;
    }

    return dummyNode.next;
};

/**
 * Definition for singly-linked list with a random pointer.
 * function RandomListNode(label) {
 *     this.label = label;
 *     this.next = this.random = null;
 * }
 */

/**
 * @param {RandomListNode} head
 * @return {RandomListNode}
 */
var copyRandomListII = function(head) {

    if ( head == null ) {
        return head;
    }

    var map = new Map();
    var newHead = new RandomListNode(head.label);
    map.set(head, newHead);
    var pre = newHead;
    var node = head.next;

    while ( node != null ) {
        var newNode = new RandomListNode(node.label);
        map.set(node, newNode);
        pre.next = newNode;
        pre = newNode;
        node = node.next;
    }

    node = head;
    var copyNode = newHead;

    while ( node != null ) {
        if ( node.random != null ) {
            copyNode.random = map.get(node.random);
        }
        copyNode = copyNode.next;
        node = node.next;
    }

    return newHead;
};

File: 138. Copy List with Random Pointer.py
# Definition for singly-linked list with a random pointer.
class RandomListNode(object):
    def __init__(self, x):
        self.label = x
        self.next = None
        self.random = None

class Solution(object):
    def copyRandomList(self, head):
        """
        :type head: RandomListNode
        :rtype: RandomListNode
        """
        if head is None:
            return None
        hash = {}
        chead = RandomListNode(head.label)
        hash[head] = chead
        p = head
        q = chead
        # 建立next的映射
        while p:
            q.random = p.random
            if p.next:
                q.next = RandomListNode(p.next.label)
                hash[p.next] = q.next
            else:
                q.next = None
            p = p.next
            q = q.next
        
        # 建立random的映射
        p = chead
        while p:
            if p.random:
                p.random = hash[p.random]
            p = p.next
        return chead
                